File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.py
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.k=k
        self.A=[]
        for i in nums:
            if len(self.A)<k:
                heappush(self.A,i)
            else:
                if self.A[0]<i:
                    heappop(self.A)
                    heappush(self.A,i)
        

    def add(self, val: int) -> int:
        if len(self.A)<self.k:
            heappush(self.A,val)
        elif self.A[0]<val:
            heappop(self.A)
            heappush(self.A,val)
        return self.A[0]
        
        


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cpp
class KthLargest {
    priority_queue<int,vector<int>,greater<int>>Q;
    int pos;
public:
    KthLargest(int k, vector<int>& nums) {
        pos=k;
        for(auto x:nums){
            if(Q.size()<k)Q.push(x);
            else
            {
                if(x>Q.top())
                {
                    Q.pop();
                    Q.push(x);
                }
            }
        }
    }
    
    int add(int val) {
        if(Q.size()>pos and val<Q.top())return Q.top();
        Q.push(val);
        if(Q.size()>pos)Q.pop();
        return Q.top();
        
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */